### ISSUE TITLE
torchaudio.transforms.Spectrogram with power=1 can easily NaN

### ISSUE THREAD

**Issue by gibiansky** [pytorch/audio#993]
## 🐛 Bug

If you use STFT through Spectrogram, with power=1, you can easily get a NaN. If your spectrogram square magnitude underflows to zero, you end up taking backprop of sqrt(0), which leads to a NaN. 

## To Reproduce

1. Use torchaudio.transforms.Spectrogram, set power=1, do a backward pass.

## Expected behavior

There should be a configurable epsilon to clamp the input to sqrt, otherwise users have to use torch.stft directly or use other workarounds (such as power=2).

**Comment by mthrok** [comment]
Hi @gibiansky 

Thanks for the report. I think I got the gist, but could you illustrate the fix?

Is it here?

https://github.com/pytorch/audio/blob/0076ab073d1ee6160efbc239e075196b35ed850b/torchaudio/functional.py#L95

**Comment by gibiansky** [comment]
Not quite. 

Here is the issue. Using torchaudio.transforms.Spectrogram with power=1 is equivalent to:

```
stft = torch.sftf(waveform, n_fft, hop_length, win_length, ...)
norm_sq = stft.pow(2.).sum(-1)
result = norm_sq.sqrt()
```

Let's suppose your waveform is zero, and so your norm_sq is zero. Then you take the square root of that, and you get zero. _However_, during the backward pass, the sqrt of zero is mathematically undefined at one, and torch.sqrt gradient is NaN.

I want to be able to do this:

```
stft = torch.sftf(waveform, n_fft, hop_length, win_length, ...)
norm_sq = stft.pow(2.).sum(-1)
result = (1e-9 + norm_sq).sqrt()
```

This ensures that we cannot get NaN on backward pass.

Here is a related issue also caused by sqrt of zero backward pass: 

https://github.com/pytorch/pytorch/issues/43211
https://github.com/pytorch/pytorch/issues/37323

I think perhaps this has been fixed in `norm()`, but computing the norm by hand still causes the issue, because it wasn't fixed for `sqrt`:

https://github.com/pytorch/pytorch/blob/201ba139115adeecae4f094a9c9790200e53ff99/tools/autograd/templates/Functions.cpp#L115

**Comment by gibiansky** [comment]
In fact, the issue is caused by the line [below the one you linked](https://github.com/pytorch/audio/blob/0076ab073d1ee6160efbc239e075196b35ed850b/torchaudio/functional.py#L97), @mthrok. Due to the [`torch.norm` bug linked in the code](https://github.com/pytorch/pytorch/issues/34279), you implemented `complex_norm`, but complex_norm uses sqrt directly, and so it has poor backward pass behaviour (NaN) when the input is zero (unlike torch.norm, which special-cases zero inputs). 

With regards to fixing it, I can see that it might be a bit tough; the core issue is torch.sqrt and torch.norm. torch.norm behaves well on backward pass but is slow on CPU compared to implementing norm yourself. But the implemented complex_norm returns NaN instead of a zero subgradient at input zero, which causes problems.

My original suggestion, adding a tiny epsilon, isn't great either, because it's not really doing the right mathematical operation anymore and can fail for float16 (as in one of the things I linked).

I would recommend (a) switching complex_norm to check the device, and use torch.norm if we're on GPU and (b) changing complex_norm to torch.norm ASAP.

**Comment by mthrok** [comment]
@gibiansky Thanks for the detailed explanation.

@anjali411 If we migrate to ComplexTensor type, can the backward pass of `abs()` handle zero <-> NaN issue mentioned above??

**Comment by anjali411** [comment]
> @anjali411 If we migrate to ComplexTensor type, can the backward pass of `abs()` handle zero <-> NaN issue mentioned above??
yeah here's a sample code to show that:
```
>>> x=torch.zeros(2, 3, dtype=torch.cfloat, requires_grad=True)
>>> y=x.abs()
>>> y.sum().backward()
>>> x.grad
tensor([[0.+0.j, 0.+0.j, 0.+0.j],
        [0.+0.j, 0.+0.j, 0.+0.j]])

>>> x1=torch.zeros(2, 3, dtype=torch.float, requires_grad=True)
>>> y1=x1.sqrt()
>>> y1.sum().backward()
>>> x1.grad
tensor([[inf, inf, inf],
        [inf, inf, inf]])
```

**Comment by mthrok** [comment]
@anjali411 Thanks!

@gibiansky Do you think this approach is sufficient for the issue you are facing?

**Comment by gibiansky** [comment]
I am unfortunately somewhat new to PyTorch (I have been using it for all of three days at this point) so I may be missing subtleties, but at least on the surface level it seems like this would solve my issue neatly.

**Comment by mthrok** [comment]
@gibiansky I made a PR #1013 . Could you take a look at the test to see if it is solving the issue? I checked that the test fails before applying the patch, but passes after.

**Comment by gibiansky** [comment]
Yes, I believe that is correct. Thank you!

**Comment by JXuann** [comment]
> Not quite.
> 
> Here is the issue. Using torchaudio.transforms.Spectrogram with power=1 is equivalent to:
> 
> ```
> stft = torch.sftf(waveform, n_fft, hop_length, win_length, ...)
> norm_sq = stft.pow(2.).sum(-1)
> result = norm_sq.sqrt()
> ```
> 
> Let's suppose your waveform is zero, and so your norm_sq is zero. Then you take the square root of that, and you get zero. _However_, during the backward pass, the sqrt of zero is mathematically undefined at one, and torch.sqrt gradient is NaN.
> 
> I want to be able to do this:
> 
> ```
> stft = torch.sftf(waveform, n_fft, hop_length, win_length, ...)
> norm_sq = stft.pow(2.).sum(-1)
> result = (1e-9 + norm_sq).sqrt()
> ```
> 
> This ensures that we cannot get NaN on backward pass.
> 
> Here is a related issue also caused by sqrt of zero backward pass:
> 
> [pytorch/pytorch#43211](https://github.com/pytorch/pytorch/issues/43211) [pytorch/pytorch#37323](https://github.com/pytorch/pytorch/issues/37323)
> 
> I think perhaps this has been fixed in `norm()`, but computing the norm by hand still causes the issue, because it wasn't fixed for `sqrt`:
> 
> https://github.com/pytorch/pytorch/blob/201ba139115adeecae4f094a9c9790200e53ff99/tools/autograd/templates/Functions.cpp#L115

@gibiansky could you briefly explain what is this for? Many thanks in advance!
```
norm_sq = stft.pow(2.).sum(-1)
result = norm_sq.sqrt()
```
